nexus/src/app/background/plan_execution.rs

```diff
@@ -0,0 +1,408 @@
+use nexus_db_queries::db::lookup::LookupPath;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::{Blueprint, OmicronZonesConfig};
+use serde_json::json;
+use sled_agent_client::Client as SledAgentClient;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Background task that takes a [`Blueprint`] and realizes the change to
+/// the state of the system based on the `Blueprint`.
+pub struct PlanExecutor {
+    datastore: Arc<DataStore>,
+    rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+}
+
+impl PlanExecutor {
+    // Temporary until we wire up the background task
+    #[allow(unused)]
+    pub fn new(
+        datastore: Arc<DataStore>,
+        rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+    ) -> PlanExecutor {
+        PlanExecutor { datastore, rx_blueprint }
+    }
+
+    // This is a modified copy of the functionality from `nexus/src/app/sled.rs`.
+    // There's no good way to access this functionality right now since it is a
+    // method on the `Nexus` type. We want to have a more constrained type we can
+    // pass into background tasks for this type of functionality, but for now we
+    // just copy the functionality.
+    async fn sled_client(
+        &self,
+        opctx: &OpContext,
+        sled_id: &Uuid,
+    ) -> Result<SledAgentClient, anyhow::Error> {
+        let (.., sled) = LookupPath::new(opctx, &self.datastore)
+            .sled_id(*sled_id)
+            .fetch()
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to create sled_agent::Client for sled_id: {}",
+                    sled_id
+                )
+            })?;
+        let dur = std::time::Duration::from_secs(60);
+        let client = reqwest::ClientBuilder::new()
+            .connect_timeout(dur)
+            .timeout(dur)
+            .build()
+            .unwrap();
+        Ok(SledAgentClient::new_with_client(
+            &format!("http://{}", sled.address()),
+            client,
+            opctx.log.clone(),
+        ))
+    }
+
+    async fn realize_blueprint(
+        &self,
+        opctx: &OpContext,
+        blueprint: &std::sync::Arc<Blueprint>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let log = opctx.log.new(o!("comment" => blueprint.comment.clone()));
+        self.deploy_zones(&log, opctx, &blueprint.omicron_zones).await
+    }
+
+    async fn deploy_zones(
+        &self,
+        log: &Logger,
+        opctx: &OpContext,
+        zones: &BTreeMap<Uuid, OmicronZonesConfig>,
+    ) -> Result<(), Vec<anyhow::Error>> {
+        let errors: Vec<_> = stream::iter(zones.clone())
+            .filter_map(|(sled_id, config)| async move {
+                let client = match self.sled_client(&opctx, &sled_id).await {
+                    Ok(client) => client,
+                    Err(err) => {
+                        warn!(log, "{}", err);
```

Comment:
Nit: probably want 

```suggestion
                        warn!(log, "{err:#}");
```

to get the full error chain.

Reply:
Fixed in 24b34f70f76a5795d51987442f392dc374cc71eb